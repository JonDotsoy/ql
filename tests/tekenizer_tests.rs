#[macro_use]
extern crate assert_matches;

#[cfg(test)]
mod tekenizer_tests {

    use insta::assert_debug_snapshot;
    use ql::Tokenizer;

    #[test]
    fn tekenizer() {
        let payload = r#"   SELECT  * FROM cars   "#;

        let tokens = Tokenizer::lexer(payload);

        assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_2() {
        let payload = r#"SELECT"#;

        let tokens = Tokenizer::lexer(payload);

        assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_string() {
        let payload = r#" "any world" try"#;

        let ref tokens = Tokenizer::lexer(payload);

        assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_string_scape() {
        let payload = r#" "any \'world" "#;

        let ref tokens = Tokenizer::lexer(payload);

        assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_numeric() {
        let payload = r#"9"#;

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_numeric_2() {
        let payload = r#"9.12"#;

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_numeric_3() {
        let payload = r#"9_123_123.122_123_943"#;

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_parenthesis_and_bracket() {
        let payload = r#"(var_1)[var_2]"#;

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_newline() {
        let payload = "abc\ndef";

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_newline_2() {
        let payload = "abc\r\ndef";

        let ref tokens = Tokenizer::lexer(payload);

        // assert_matches!(tokens, Ok(_));
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tekenizer_lexer_query_1() {
        let payload = "ctx.var_name:123 ";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_sample_1() {
        let payload = "search value";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_sample_2() {
        let payload = "search value condition:";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_sample_3() {
        let payload = r#"search value condition: "value""#;
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_sample_4() {
        let payload = r#"
            search value
            condition: "value"
            context.condition > "value"
        "#;
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_string_with_scape_char() {
        let payload = r#""i'm string \"""#;
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_template_1() {
        let payload = "`abc${3}def`";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_template_2() {
        let payload = "`abc\\${3}def`";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_string_template_1() {
        let payload = "`a${`b${\"c\"}d`}e`";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_string_template_2() {
        let payload = "`a$\\`b$\"c\"}d\\`}e`";
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_string_template_3() {
        let payload = r#"`first string ${keyword} second string ${12_20 + `${"hola"}`}`"#;
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_string_template_4() {
        let payload = r#"`a${b{c}d}e`"#;
        let ref tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_selector_with_variables() {
        let payload = r#"
            selector1.[$variable1].name
        "#;
        let tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }

    #[test]
    fn tokenizer_lexer_selector_with_variables_2() {
        let payload = r#"
            $.user = $VAR && (
                $.tag[$TAG_NAME] = true ||
                $.tag[`${TAG_NAME}_alt`] = true
            )
        "#;
        let tokens = Tokenizer::lexer(payload);
        assert_debug_snapshot!(tokens);
    }
}
