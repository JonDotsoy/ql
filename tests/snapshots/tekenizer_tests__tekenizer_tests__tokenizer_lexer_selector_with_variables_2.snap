---
source: tests/tekenizer_tests.rs
expression: tokens
---
Ok(
    [
        Token {
            kind: "newline",
            raw: "\n",
            span: Span {
                start: 0,
                end: 1,
            },
        },
        Token {
            kind: "keyword",
            raw: "$",
            span: Span {
                start: 13,
                end: 14,
            },
        },
        Token {
            kind: "dot",
            raw: ".",
            span: Span {
                start: 14,
                end: 15,
            },
        },
        Token {
            kind: "keyword",
            raw: "user",
            span: Span {
                start: 15,
                end: 19,
            },
        },
        Token {
            kind: "equal",
            raw: "=",
            span: Span {
                start: 20,
                end: 21,
            },
        },
        Token {
            kind: "keyword",
            raw: "$VAR",
            span: Span {
                start: 22,
                end: 26,
            },
        },
        Token {
            kind: "operation",
            raw: "&&",
            span: Span {
                start: 27,
                end: 29,
            },
        },
        Token {
            kind: "open_curly_bracket",
            raw: "(",
            span: Span {
                start: 30,
                end: 31,
            },
        },
        Token {
            kind: "newline",
            raw: "\n",
            span: Span {
                start: 31,
                end: 32,
            },
        },
        Token {
            kind: "keyword",
            raw: "$",
            span: Span {
                start: 48,
                end: 49,
            },
        },
        Token {
            kind: "dot",
            raw: ".",
            span: Span {
                start: 49,
                end: 50,
            },
        },
        Token {
            kind: "keyword",
            raw: "tag",
            span: Span {
                start: 50,
                end: 53,
            },
        },
        Token {
            kind: "open_square_bracket",
            raw: "[",
            span: Span {
                start: 53,
                end: 54,
            },
        },
        Token {
            kind: "keyword",
            raw: "$TAG_NAME",
            span: Span {
                start: 54,
                end: 63,
            },
        },
        Token {
            kind: "close_square_bracket",
            raw: "]",
            span: Span {
                start: 63,
                end: 64,
            },
        },
        Token {
            kind: "equal",
            raw: "=",
            span: Span {
                start: 65,
                end: 66,
            },
        },
        Token {
            kind: "keyword",
            raw: "true",
            span: Span {
                start: 67,
                end: 71,
            },
        },
        Token {
            kind: "operation",
            raw: "||",
            span: Span {
                start: 72,
                end: 74,
            },
        },
        Token {
            kind: "newline",
            raw: "\n",
            span: Span {
                start: 74,
                end: 75,
            },
        },
        Token {
            kind: "keyword",
            raw: "$",
            span: Span {
                start: 91,
                end: 92,
            },
        },
        Token {
            kind: "dot",
            raw: ".",
            span: Span {
                start: 92,
                end: 93,
            },
        },
        Token {
            kind: "keyword",
            raw: "tag",
            span: Span {
                start: 93,
                end: 96,
            },
        },
        Token {
            kind: "open_square_bracket",
            raw: "[",
            span: Span {
                start: 96,
                end: 97,
            },
        },
        Token {
            kind: "template_start",
            raw: "`",
            span: Span {
                start: 97,
                end: 98,
            },
        },
        Token {
            kind: "template",
            raw: "",
            span: Span {
                start: 98,
                end: 98,
            },
        },
        Token {
            kind: "template_start_expression",
            raw: "${",
            span: Span {
                start: 98,
                end: 100,
            },
        },
        Token {
            kind: "keyword",
            raw: "TAG_NAME",
            span: Span {
                start: 100,
                end: 108,
            },
        },
        Token {
            kind: "template_close_expression",
            raw: "}",
            span: Span {
                start: 108,
                end: 109,
            },
        },
        Token {
            kind: "template",
            raw: "_alt",
            span: Span {
                start: 109,
                end: 113,
            },
        },
        Token {
            kind: "template_close",
            raw: "`",
            span: Span {
                start: 113,
                end: 114,
            },
        },
        Token {
            kind: "close_square_bracket",
            raw: "]",
            span: Span {
                start: 114,
                end: 115,
            },
        },
        Token {
            kind: "equal",
            raw: "=",
            span: Span {
                start: 116,
                end: 117,
            },
        },
        Token {
            kind: "keyword",
            raw: "true",
            span: Span {
                start: 118,
                end: 122,
            },
        },
        Token {
            kind: "newline",
            raw: "\n",
            span: Span {
                start: 122,
                end: 123,
            },
        },
        Token {
            kind: "close_parenthesis",
            raw: ")",
            span: Span {
                start: 135,
                end: 136,
            },
        },
        Token {
            kind: "newline",
            raw: "\n",
            span: Span {
                start: 136,
                end: 137,
            },
        },
    ],
)
